src/slice.rs

```diff
@@ -240,220 +260,342 @@ impl From<RangeFull> for Slice {
     }
 }
 
-impl From<RangeFull> for SliceOrIndex {
+impl From<RangeFull> for AxisSliceInfo {
     #[inline]
-    fn from(_: RangeFull) -> SliceOrIndex {
-        SliceOrIndex::Slice {
+    fn from(_: RangeFull) -> AxisSliceInfo {
+        AxisSliceInfo::Slice {
             start: 0,
             end: None,
             step: 1,
         }
     }
 }
 
-impl From<Slice> for SliceOrIndex {
+impl From<Slice> for AxisSliceInfo {
     #[inline]
-    fn from(s: Slice) -> SliceOrIndex {
-        SliceOrIndex::Slice {
+    fn from(s: Slice) -> AxisSliceInfo {
+        AxisSliceInfo::Slice {
             start: s.start,
             end: s.end,
             step: s.step,
         }
     }
 }
 
-macro_rules! impl_sliceorindex_from_index {
+macro_rules! impl_axissliceinfo_from_index {
     ($index:ty) => {
-        impl From<$index> for SliceOrIndex {
+        impl From<$index> for AxisSliceInfo {
             #[inline]
-            fn from(r: $index) -> SliceOrIndex {
-                SliceOrIndex::Index(r as isize)
+            fn from(r: $index) -> AxisSliceInfo {
+                AxisSliceInfo::Index(r as isize)
+            }
+        }
+    };
+}
+impl_axissliceinfo_from_index!(isize);
+impl_axissliceinfo_from_index!(usize);
+impl_axissliceinfo_from_index!(i32);
+
+impl From<NewAxis> for AxisSliceInfo {
+    #[inline]
+    fn from(_: NewAxis) -> AxisSliceInfo {
+        AxisSliceInfo::NewAxis
+    }
+}
+
+/// A type that can slice an array of dimension `D`.
+///
+/// This trait is unsafe to implement because the implementation must ensure
+/// that `D`, `Self::OutDim`, `self.in_dim()`, and `self.out_ndim()` are
+/// consistent with the `&[AxisSliceInfo]` returned by `self.as_ref()`.
+pub unsafe trait CanSlice<D: Dimension>: AsRef<[AxisSliceInfo]> {
```

Comment:
should probably use the private trait marker because we don't want it as an extension surface (or would that be useful?)

Reply:
Okay, I've added the private marker to `CanSlice` and `MultiSlice` and exposed them in the public API.